File: LinkedList/C/Insertion/insertafterKEY.c
#include<stdio.h>
#include<stdlib.h>

struct node {
    int data;
    struct node *next;
}nn;

struct node *insertend(struct node *head,int data){
    struct node *newnode=(struct node*)malloc(sizeof(nn));
    newnode->data=data;
    newnode->next=NULL;
    struct node*ptr=head;
    while(ptr->next != NULL){
        ptr=ptr->next;
    }
    ptr->next=newnode;
    return head;
}

void printnode(struct node *head){
    struct node *ptr = head;
    while(ptr != NULL) {
        printf("%d -> ", ptr->data);
        ptr = ptr->next;
    }
    printf("NULL\n");
}

// Insert after key
struct node *insertafterkey(struct node *head, int data, int key) {

    struct node *newnode = malloc(sizeof(struct node));
    newnode->data = data;
    newnode->next = NULL;
    struct node *ptr = head;
    // Find the key
    while (ptr != NULL && ptr->data != key) {
        ptr = ptr->next;
    }

    // Key not found
    if (ptr == NULL) {
        printf("Key not found.\n");
        free(newnode);
        return head;
    }
    // Insert after the key
    newnode->next = ptr->next;
    ptr->next = newnode;
    return head;
}

int main(){
    int data,key;
    struct node *head = (struct node*)malloc(sizeof(struct node));
    head->data = 1;
    head->next = NULL;

    head = insertend(head, 2);
    head = insertend(head, 3);
    head = insertend(head, 4);
    head = insertend(head, 5);
    printnode(head);

    printf("Enter the data to be inserted and the key[element]: ");
    scanf("%d %d",&data,&key);
    head=insertafterkey(head,data,key);
    printnode(head);
    
    return 0;
}

File: LinkedList/Python/Insertion/insertafterKEY.py
class node:
    def __init__(self,data):
        self.data=data
        self.next=None

class LinkedList:
    def __init__(self):
        self.head=None

    def insertEND(self,data):
        newnode=node(data)
        ptr=self.head
        if ptr == None:
            self.head=newnode
        else:
            while(ptr.next != None):
                ptr=ptr.next
            ptr.next=newnode

    def display(self):
        ptr=self.head
        while ptr is not None:
            print(ptr.data, end=" -> ")
            ptr=ptr.next
        print("NULL")

    def insertafterkey(self, data, key):
        newnode = node(data)
        ptr = self.head
        # Case 1: Head is NULL
        if ptr == None:
            self.head = newnode
        else:
            prev = None
            # Search for key
            while ptr != None and ptr.data != key:
                prev = ptr
                ptr = ptr.next
            if ptr != None:
                newnode.next = ptr.next
                ptr.next=newnode
            # Case 4: Key not found
            else:
                print("Key not found.")
        

head=LinkedList()
head.insertEND(1);
head.insertEND(2);
head.insertEND(3);
head.insertEND(4);
head.insertEND(5);
head.display()
data,key=map(int,input("Enter the data to be inserted and the key[element] : ").split())
head.insertafterkey(data,key)
head.display()
